lib/phoenix/channel/transport.ex

```diff
@@ -70,39 +85,53 @@ defmodule Phoenix.Channel.Transport do
 
   The server will respond to heartbeats with the same message
   """
-  def dispatch(_, "phoenix", "heartbeat", _payload, transport_pid, _router, _pubsub_server, _transport) do
+  def dispatch(_, %{topic: "phoenix", event: "heartbeat"}, transport_pid, _router, _pubsub_server, _transport) do
     send transport_pid, {:socket_push, %Message{topic: "phoenix", event: "heartbeat", payload: %{}}}
   end
-  def dispatch(nil, topic, "join", payload, transport_pid, router, endpoint, transport) do
-    case router.channel_for_topic(topic, transport) do
-      nil     -> log_ignore(topic, router)
+  def dispatch(nil, %{event: "phx_join"} = msg, transport_pid, router, endpoint, transport) do
+    case router.channel_for_topic(msg.topic, transport) do
+      nil     -> log_ignore(msg.topic, router)
       channel ->
         socket = %Socket{transport_pid: transport_pid,
                   router: router,
                   endpoint: endpoint,
                   pubsub_server: endpoint.__pubsub_server__(),
-                  topic: topic,
+                  topic: msg.topic,
+                  ref: msg.ref,
                   channel: channel,
                   transport: transport}
 
-        Phoenix.Channel.Server.start_link(socket, payload)
+        Phoenix.Channel.Server.start_link(socket, msg.payload)
     end
   end
-  def dispatch(nil, topic, _event, _payload, _adapter_pid, router, _pubsub_server, _transport) do
-    log_ignore(topic, router)
+  def dispatch(nil, msg, _transport_pid, router, _pubsub_server, _transport) do
+    log_ignore(msg.topic, router)
     :ignore
   end
-  def dispatch(socket_pid, _topic, event, payload, _adapter_pid, _router, _pubsub_server, _transport) do
-    GenServer.cast(socket_pid, {:handle_in, event, payload})
+  def dispatch(socket_pid, msg, _transport_pid, _router, _pubsub_server, _transport) do
+    GenServer.cast(socket_pid, {:handle_in, msg.event, msg.payload, msg.ref})
     :ok
   end
-
   defp log_ignore(topic, router) do
     Logger.debug fn -> "Ignoring unmatched topic \"#{topic}\" in #{inspect(router)}" end
     :ignore
   end
 
   @doc """
+  Returns the `%Phoenix.Message{}` for a channel close event
+  """
+  def chan_close_message(topic) do
+    %Message{topic: topic, event: "phx_chan_close", payload: %{}}
+  end
+
+  @doc """
+  Returns the `%Phoenix.Message{}` for a channel error event
+  """
+  def chan_error_message(topic) do
+    %Message{topic: topic, event: "phx_chan_error", payload: %{}}
```

Comment:
Sorry, minor nitpicking. You have used `phx_join` up and `phx_chan_close` error. Should we unify them?